app/src/test/java/org/sagebionetworks/bridge/researchstack/CrfDataProviderTest.java

```diff
@@ -0,0 +1,514 @@
+        PowerMockito.mockStatic(Looper.class);
+
+        AppPrefs.init(context);
+    }
+
+    private void resetTest(DateTime firstSignInDate, boolean assignClinic1, String initialDataGroup) {
+        dataProvider = new MockCrfDataProvider(
+                firstSignInDate, assignClinic1, initialDataGroup,
+                researchStackDAO, storageAccess, taskHelper);
+        dataProvider.mockGetAllActiities(successList(assignClinic1));
+    }
+
+    @Test
+    public void testInitialize() {
+        dataProvider.initialize(context).test().assertCompleted();
+    }
+
+    @Test
+    public void testFindClinc1() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic1");
+        assertNotNull(clinic);
+        assertEquals("clinic1", clinic.getActivity().getTask().getIdentifier());
+    }
+
+    @Test
+    public void testFindClinc2() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic2");
+        assertNotNull(clinic);
+        assertEquals("clinic2", clinic.getActivity().getTask().getIdentifier());
+    }
+
+    @Test
+    public void testFindFail() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic3");
+        assertNull(clinic);
+    }
+
+    @Test
+    public void testAssignClinic1() {
+        resetTest(null, true, null);
+        final ScheduledActivityListV4 successList = successList(true);
+        dataProvider.mockGetAllActiities(successList);
+
+        dataProvider.mockGetClinicActiities(clinicActivities(null, null));
+        dataProvider.getCrfActivities(new CrfDataProvider.CrfActivitiesListener() {
+            @Override
+            public void success(ScheduledActivityListV4 activityList) {
+                assertEquals(activityList, successList);
+                assertEquals(dataProvider.getCrfPrefs().getFirstSignInDate(), dataProvider.COMPLETION_TIME_CLINIC1);
+            }
+
+            @Override
+            public void error(String localizedError) {
+                assertFalse(true);  // fail
+            }
+        });
+    }
+
+    @Test
+    public void testAssignClinic2() {
+        resetTest(null, false, null);
+        final ScheduledActivityListV4 successList = successList(false);
+        dataProvider.mockGetAllActiities(successList);
+
+        dataProvider.mockGetClinicActiities(clinicActivities(null, null));
+        dataProvider.getCrfActivities(new CrfDataProvider.CrfActivitiesListener() {
+            @Override
+            public void success(ScheduledActivityListV4 activityList) {
+                assertEquals(activityList, successList);
+                assertEquals(dataProvider.getCrfPrefs().getFirstSignInDate(), dataProvider.COMPLETION_TIME_CLINIC2);
```

Comment:
Maybe instead of duplicating this in every test, it would be better to create a subclass TestCrfActivitiesListener, which takes in the expected activityList and expected firstSignInDate in the constructor. This way, you can declare the class once and use it everywhere.